server/api.go

```diff
@@ -645,8 +625,127 @@ func (a *API) getConnectedUsersFile(w http.ResponseWriter, r *http.Request) {
+		a.p.API.LogWarn("Unable to write headers in CSV file", "error", err.Error())
+		http.Error(w, "unable to write data in CSV file", http.StatusInternalServerError)
+		return
+	}
+
+	for _, email := range whitelist {
+		if err := csvWriter.Write([]string{email}); err != nil {
+			a.p.API.LogWarn("Unable to write data in CSV file", "error", err.Error())
+			http.Error(w, "unable to write data in CSV file", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	csvWriter.Flush()
+	if err := csvWriter.Error(); err != nil {
+		a.p.API.LogWarn("Unable to flush the data in writer", "error", err.Error())
+		http.Error(w, "unable to write data in CSV file", http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/csv")
+	w.Header().Set("Content-Disposition", "attachment;filename=invite-whitelist.csv")
+	if _, err := w.Write(b.Bytes()); err != nil {
+		a.p.API.LogWarn("Unable to write the data", "error", err.Error())
+		http.Error(w, "unable to write the data", http.StatusInternalServerError)
+	}
+}
+
+func (a *API) updateWhitelist(w http.ResponseWriter, r *http.Request) {
+	userID := r.Header.Get("Mattermost-User-ID")
+	if userID == "" {
+		a.p.API.LogWarn("Not authorized")
+		http.Error(w, "not authorized", http.StatusUnauthorized)
+		return
+	}
+
+	if !a.p.API.HasPermissionTo(userID, model.PermissionManageSystem) {
+		a.p.API.LogWarn("Insufficient permissions", "user_id", userID)
+		http.Error(w, "not able to authorize the user", http.StatusForbidden)
+		return
+	}
+
+	file, _, err := r.FormFile("file")
+	if err != nil {
+		a.p.API.LogWarn("Error reading whitelist file")
+		http.Error(w, "error reading whitelist", http.StatusBadRequest)
+		return
+	}
+
+	reader := csv.NewReader(file)
+	columns, err := reader.Read()
+	if err != nil || strings.ToLower(columns[0]) != "email" || len(columns) != 1 {
+		a.p.API.LogWarn("Error parsing whitelist csv header")
+		http.Error(w, "error parsing whitelist - bad csv header", http.StatusBadRequest)
+		return
+	}
+
+	records, err := reader.ReadAll()
+	if err != nil {
+		a.p.API.LogWarn("Error parsing whitelist csv data")
+		http.Error(w, "error parsing whitelist", http.StatusBadRequest)
+	}
+
+	var users []*model.User
+
+	for _, row := range records {
+		email := row[0]
+
+		user, err := a.p.API.GetUserByEmail(email)
+		if err != nil {
+			a.p.API.LogWarn("Error getting user")
+			http.Error(w, "error processing whitelist - could not find user with email: "+email, http.StatusInternalServerError)
+			return
+		}
+
+		users = append(users, user)
+	}
+
+	if err := a.p.store.DeleteWhitelist(); err != nil {
+		a.p.API.LogWarn("Error deleting whitelist")
+		http.Error(w, "error processing whitelist", http.StatusInternalServerError)
+		return
+	}
+
+	for _, user := range users {
+		if err := a.p.store.StoreUserInWhitelist(user.Id); err != nil {
+			a.p.API.LogWarn("Error adding user to whitelist")
+			http.Error(w, "error processing whitelist", http.StatusInternalServerError)
+			return
+		}
+	}
```

Comment:
I'd suggest doing everything in one go, both deleting the current whitelist and adding all users, so we treat the operation as atomic. We could have a `a.p.store.ReplaceWhitelist(users)` and inside of it, open a `tx`, delete the current whitelist and run a single query that inserts all users before committing the `tx`.

As we cannot add a single user to the whitelist, in case of failure, the user will send the whole whitelist fixed and it will be processed again, deleting the intermediate state that was kept before, so from my PoV it makes sense to think of this as an atomic operation that either works fully or returns an error and changes nothing rather than an intermediate state that will be overwritten in the next attempt.

What we could do to improve the feedback loop on the CSV is to iterate through the whole file, store those users that failed and return an error for all of them, so they can be fixed in one go rather than in several requests:

```go
	var users []*model.User
	var notFoundUsers []string
	for _, row := range records {
		email := row[0]

		user, err := a.p.API.GetUserByEmail(email)
		if err != nil {
			a.p.API.LogWarn("Error getting user", "email", email)
			notFoundUsers = append(notFoundUsers, email)
		}

		users = append(users, user)
	}

	if len(notFoundUsers) != 0 {
		http.Error(w, "error processing whitelist - could not find user with email: "+strings.Join(notFoundUsers, ", "), http.StatusInternalServerError)
		return
	}
```

Reply:
This has been re-written and split better between this method which now only does the CSV stuff, and the new store method `SetWhitelist` which has a transaction and does batching and threshold=10 error reporting.

Reply:
Discussed offline, we will keep the thresholds for users not found and CSV parse errors at the API level, and the store operation will be atomic